### PR TITLE
fix: remove default fields on startup, simplify startup logic for updating global variables in LF

### LINKED ISSUE #2261: [Bug]: Unable to run Chat via OpenAI model provider when a watsonx.ai model provider was previously configured but removed
### OpenRAG Version

0.6.0 + latest from **main** (152225a148f14622edfea762a073697078775332)

### Deployment Method

Local development (make dev)

### Operating System

Ubuntu 24.04.4 LTS

### Python Version

Python 3.13.14

### Affected Area

Backend + Langflow

### Bug Description

Unable to run Chat via OpenAI model provider when a watsonx.ai model provider was previously configured but removed

### Steps to Reproduce

- Use a workspace with a previously configured watsonx.ai model provider; but, is currently removed / unconfigured
   - i.e. No factory reset
   - Note: watsonx env vars are still specified in `.env` file
   - Note: Using latest version of Langflow
   - Note: The workspace upgraded the version of Langflow from an old version when started
- Configure OpenAI model provider
   - Make sure OpenAI language and embedding model provider is selected in Settings
- Enter a sample prompt in Chat
   - BUG: Chat error shown indicating that the watsonx model provider API key is incorrect - not the OpenAI API key
   - WORKAROUND
      - Open Langflow
      - Go to Settings
      - Manually delete the 3 "WATSONX*" env vars 
      - Restart Backend application

### Expected Behavior

Chat should work with OpenAI configured model provider

### Actual Behavior

Chat does not work and is trying to use an unconfigured watsonx.ai model provider

### Relevant Logs

- Was seeing a 401 Unauthenticated error in the logs for the Backend application; but, didn't collect the logs

```shell
N/A
```

### Screenshots

Image

Image

### Additional Context

- N/A

### Checklist

- [x] I have searched existing issues to ensure this bug hasn't been reported before.
- [x] I have provided all the requested information.

### PR DESCRIPTION
**Pull request by lucaseduoli**
This pull request improves the handling and synchronization of Langflow global variables to ensure all plain-string global variables are of type `Generic` and that any `default_fields` ("Apply To" fields) are removed for backward compatibility. It also updates the update/create logic to consistently clear `default_fields` and adds a new test to verify this behavior.

**Langflow global variable synchronization and migration:**

* Enhanced the `ensure_required_langflow_global_variables` function in `langflow_sync.py` to update existing variables so that all plain-string globals are set to type `Generic` and to remove any `default_fields` from all variables, ensuring backward compatibility and consistency.
* Updated the `_update_langflow_global_variable` method in `settings.py` to always clear `default_fields` when updating or recreating variables, and to skip updates only if both the value and `default_fields` are already correct. [[1]](diffhunk://#diff-6ff8f6f489c49cfc2185aa6bdee4b6ee87be730175eed7d6a67fc11194861690R1546-R1550) [[2]](diffhunk://#diff-6ff8f6f489c49cfc2185aa6bdee4b6ee87be730175eed7d6a67fc11194861690L1561-L1563) [[3]](diffhunk://#diff-6ff8f6f489c49cfc2185aa6bdee4b6ee87be730175eed7d6a67fc11194861690L1598-R1596) [[4]](diffhunk://#diff-6ff8f6f489c49cfc2185aa6bdee4b6ee87be730175eed7d6a67fc11194861690L1624-R1623) [[5]](diffhunk://#diff-6ff8f6f489c49cfc2185aa6bdee4b6ee87be730175eed7d6a67fc11194861690L1637-R1636)

**Testing improvements:**

* Modified test fixtures in `test_langflow_global_variables.py` to expect empty `default_fields` for all variables, reflecting the new desired state. [[1]](diffhunk://#diff-a768d729e08d10aee2c7515108b726894545ac5ecb8a568c099d7b4f3627d9f0L91-R91) [[2]](diffhunk://#diff-a768d729e08d10aee2c7515108b726894545ac5ecb8a568c099d7b4f3627d9f0L136-R214)
* Added a new test, `test_ensure_required_langflow_global_variables_removes_apply_to_fields`, to verify that the synchronization logic correctly issues a PATCH request to remove `default_fields` from variables that have them.

These changes ensure that Langflow global variables are consistently managed and compatible with expected usage, and that tests accurately reflect and validate the new logic.

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved synchronization of required Langflow variables.
  * Existing values are retained while outdated default fields are cleared.
  * Variables are recreated automatically when their types change.
  * Missing required variables are created during synchronization.
  * Synchronization continues when an individual variable update, creation, deletion, or update operation fails.
  * Failed variable retrieval now safely prevents incomplete synchronization.
  * Optional configuration sections are handled safely, preventing unrelated synchronization failures.
  * Synchronization reports clear warnings for variables that could not be reconciled.

### CODE REVIEW

**Review by mpawlow — CHANGES_REQUESTED** [review]
## Functional Review 1

- ✅ Successfully reproduced the problem off of **main** in my secondary workspace
- ⚠️ Unfortunately, the changes in this PR didn't resolve the problem

### Relevant Backend Logs

```
[2026-08-19T20:21:41.229255Z] [DEBUG  ] [settings.py:831               ] get_langflow_api_key called
                                                                         - current_key_present: True
                                                                         - force_regenerate: False
[2026-08-19T20:21:41.265077Z] [INFO   ] [langflow_sync.py:170          ] Updating Langflow global variable
                                                                         - variable_name: WATSONX_APIKEY
                                                                         - variable_id: ef486950-6dc9-46b6-9070-4335ee66e6e6
[2026-08-19T20:21:41.265248Z] [DEBUG  ] [settings.py:831               ] get_langflow_api_key called
                                                                         - current_key_present: True
                                                                         - force_regenerate: False
[2026-08-19T20:21:41.297137Z] [WARNING] [settings.py:1473              ] Langflow request returned transient status, retrying
                                                                         - endpoint: /api/v1/variables/ef486950-6dc9-46b6-9070-4335ee66e6e6
                                                                         - status_code: 500
                                                                         - attempt: 1
                                                                         - max_attempts: 3
                                                                         - retry_in_seconds: 1
[2026-08-19T20:21:42.298112Z] [DEBUG  ] [settings.py:831               ] get_langflow_api_key called
                                                                         - current_key_present: True
                                                                         - force_regenerate: False
[2026-08-19T20:21:42.328521Z] [WARNING] [settings.py:1473              ] Langflow request returned transient status, retrying
                                                                         - endpoint: /api/v1/variables/ef486950-6dc9-46b6-9070-4335ee66e6e6
                                                                         - status_code: 500
                                                                         - attempt: 2
                                                                         - max_attempts: 3
                                                                         - retry_in_seconds: 2
[2026-08-19T20:21:44.330015Z] [DEBUG  ] [settings.py:831               ] get_langflow_api_key called
                                                                         - current_key_present: True
                                                                         - force_regenerate: False
[2026-08-19T20:21:44.358113Z] [INFO   ] [langflow_sync.py:154          ] Migrating Langflow global variable type
                                                                         - variable_name: OLLAMA_BASE_URL
                                                                         - old_type: Credential
                                                                         - new_type: Generic
[2026-08-19T20:21:44.358446Z] [DEBUG  ] [settings.py:831               ] get_langflow_api_key called
                                                                         - current_key_present: True
                                                                         - force_regenerate: False
[2026-08-19T20:21:44.398496Z] [DEBUG  ] [settings.py:831               ] get_langflow_api_key called
                                                                         - current_key_present: True
                                                                         - force_regenerate: False
[2026-08-19T20:21:44.439868Z] [INFO   ] [startup_orchestrator.py:220   ] Ensured required Langflow global variables
```

### Relevant Langflow Logs

(missing update global variable log events)

```
openrag-langflow  | 2026-08-19T20:39:14.272111Z [error    ] Workflow job execution failed for flow ebc01d31-1976-46ce-a385-b0240327226c: Error running graph: Error building Component OpenSearch (Multi-Model Multi-Embedding):
openrag-langflow  |
openrag-langflow  | Error code: 401 - {'error': {'message': 'Incorrect API key provided: y7x7sA6V********************************7S3h. You can find your API key at https://platform.openai.com/account/api-keys.', 'type': 'invalid_request_error', 'param': None, 'code': 'invalid_api_key'}} service=langflow
openrag-langflow  | 2026-08-19T20:39:14.273013Z [error    ] Error processing OpenAI Responses request: 500: {"message":"Error running graph: Error building Component OpenSearch (Multi-Model Multi-Embedding): \n\nError code: 401 - {'error': {'message': 'Incorrect API key provided: y7x7sA6V********************************7S3h. You can find your API key at https://platform.openai.com/account/api-keys.', 'type': 'invalid_request_error', 'param': None, 'code': 'invalid_api_key'}}","traceback":null,"description":null,"code":null,"suggestion":"The flow contains 1 outdated component. We recommend updating the following component: ChatInput-7W1BE."} service=langflow
```

### .env

```shell
# Model Provider API Keys
OPENAI_API_KEY=*********************************************************
ANTHROPIC_API_KEY=
OLLAMA_ENDPOINT=
# OLLAMA_ENDPOINT=http://9.46.110.49:11434
WATSONX_API_KEY=
# WATSONX_API_KEY=y7x7sA6V********************************7S3h
WATSONX_ENDPOINT=
# WATSONX_ENDPOINT=https://ca-tor.ml.cloud.ibm.com
WATSONX_PROJECT_ID=
# WATSONX_PROJECT_ID=45ec2f5c-c82d-4d43-9176-8e2c5d86aa68
```

### Langflow Global Variables Screenshot

Image

### OpenRAG Settings Screenshot

Image

File: src/api/settings/langflow_sync.py [hunk 1]
```diff
@@ -122,91 +122,142 @@ def _required_generic_global_values(config) -> dict[str, str]:
...
-                "WATSONX_APIKEY", config.providers.watsonx.api_key
-            )
-            logger.info("Set WATSONX_APIKEY global variable in Langflow")
 
-        if config.providers.watsonx.project_id:
-            await _upsert_langflow_global_variable(
-                "WATSONX_PROJECT_ID", config.providers.watsonx.project_id
-            )
-            logger.info("Set WATSONX_PROJECT_ID global variable in Langflow")
+    async def _safe_upsert(name: str, value: str):
```
Comment (mpawlow) [inline src/api/settings/langflow_sync.py:226]: ### (1a) [Major] `_update_langflow_global_variables` never raises, silently defeating the caller's error-propagation contract

#### Problem

- `_update_langflow_global_variables` (`src/api/settings/langflow_sync.py:200-260`) used to be wrapped in a single `try/except ... raise`, so any provider-sync failure propagated to the caller.
- The rewrite replaces that with a per-provider `_safe_upsert` helper that catches `Exception`, logs a warning, and returns — there is no `raise` anywhere left in the function body (verified: zero `raise` statements between lines 200–260).
- `src/api/settings/endpoints.py:1173-1223` still calls this function inside a `try/except Exception as e: ... raise`, expecting a failure to bubble up so the settings-save request can report an error. That `except` block is now dead code for this call — it can never fire.
- Net effect: the `POST /settings` (or equivalent settings-save) endpoint will return success even when syncing `WATSONX_APIKEY`, `OPENAI_API_KEY`, `SELECTED_LANGUAGE_MODEL_PROVIDER`, etc. into Langflow completely failed. The admin gets no indication that their provider change didn't take effect.

#### Background Information

- This is precisely the failure mode the linked issue (#2261) describes — a stale/incorrect provider variable silently persisting in Langflow while the UI reports the intended provider as configured. Turning a visible failure into a silent one increases the chance of that class of bug recurring undetected.
- The async, fire-and-forget path (`_run_async_post_save_langflow_updates`, line ~285) intentionally swallows exceptions ("Do not fail user request if async sync fails") — that's a deliberate, different call site and is fine as-is. The concern here is specifically the **synchronous** call at `endpoints.py:1194`, made before the HTTP response is returned, where callers still assume a raise-on-failure contract.

#### Code References

- `src/api/settings/langflow_sync.py:200-260` (`_update_langflow_global_variables`, `_safe_upsert`)
- `src/api/settings/endpoints.py:1173-1223` (dead `except` block)

#### Potential Solution

- Have `_safe_upsert` (and the Ollama resolution `try/except` at line 236-244) collect failures into a list and, after attempting all providers, raise a single aggregate exception (or return a failure summary) if any upsert failed — preserving "continue past a single failure" while still surfacing an overall failure to the synchronous caller.

#### Alternative Solutions

- Have `endpoints.py` inspect a returned success/failure report from `_update_langflow_global_variables` instead of relying on an exception, and surface a partial-failure warning (not necessarily a hard 500) to the admin.

File: src/api/settings/langflow_sync.py [hunk 2]
```diff
@@ -122,91 +122,142 @@ def _required_generic_global_values(config) -> dict[str, str]:
...
+
+            if curr_type != target_type:
+                logger.info(
+                    "Migrating Langflow global variable type",
+                    variable_name=name,
+                    old_type=curr_type,
+                    new_type=target_type,
+                )
+                del_resp = await clients.langflow_request("DELETE", f"/api/v1/variables/{var_id}")
+                if del_resp.status_code in (200, 204):
```
Comment (mpawlow) [inline src/api/settings/langflow_sync.py]: ### (1b) [Major] DELETE/POST/PATCH responses in the startup reconciliation loop are not status-checked — type-migration and cleanup can fail completely silently

- 👍 Concur with CodeRabbit.ai feedback

#### Problem

- In `ensure_required_langflow_global_variables` (`src/api/settings/langflow_sync.py:154-186`):
  - If `DELETE /api/v1/variables/{var_id}` (line 161) fails (not `200`/`204`), the code takes no action at all — no recreate, no log line, no exception. The mistyped variable is left in place with zero trace of the failure.
  - If the DELETE succeeds but the subsequent recreate `POST` (line 169-171) fails, the response is discarded without a status check. The variable is now **missing entirely** until the next backend startup, and nothing is logged.
  - The `PATCH` call in the `elif` branch (line 184-186) also discards its response — a failed `default_fields`/value cleanup is silent.
  - The surrounding `except Exception as e:` (line 187-188) only catches raised exceptions; since none of these paths raise on a non-2xx response, that handler never fires for these failures either.

#### Background Information

- No test in this PR exercises the type-migration (`curr_type != target_type`) branch of `ensure_required_langflow_global_variables` itself — only `_update_langflow_global_variable` in `settings.py` (a separate function) has recreate-status coverage. The gap is untested as well as unhandled.
- This function only runs once per backend startup (`services/startup_orchestrator.py:219`), so a transient Langflow hiccup here can leave a required variable broken/missing for the whole process lifetime with no operator-visible signal.

#### Potential Solution

- Check status codes explicitly and raise (letting the existing per-variable `except` at line 187 log and continue), mirroring the pattern already used in `_update_langflow_global_variable` (`src/config/settings.py:1602-1618`), which does check `recreate_response.status_code` and raises `RuntimeError` on failure:
  ```python
  del_resp = await clients.langflow_request("DELETE", f"/api/v1/variables/{var_id}")
  if del_resp.status_code not in (200, 204):
      raise RuntimeError(f"Failed to delete {name!r}: status={del_resp.status_code}")
  recreate_response = await clients.langflow_request("POST", "/api/v1/variables/", json=recreate_payload)
  if recreate_response.status_code not in (200, 201):
      raise RuntimeError(f"Failed to recreate {name!r}: status={recreate_response.status_code}")
  ```
- Add a regression test for "DELETE succeeds, POST fails" and "DELETE fails" per CodeRabbit's suggestion.

File: src/api/settings/langflow_sync.py [hunk 3]
```diff
@@ -122,91 +122,142 @@ def _required_generic_global_values(config) -> dict[str, str]:
...
+                }
+                if is_generic:
+                    patch_payload["value"] = target_val
+                await clients.langflow_request(
+                    "PATCH", f"/api/v1/variables/{var_id}", json=patch_payload
+                )
+        except Exception as e:
+            logger.warning(f"Failed to update Langflow global variable '{name}': {e}")
+
+    # Create missing generic variables
```
Comment (mpawlow) [inline src/api/settings/langflow_sync.py:210]: ### (1c) [Normal] A failed variable listing (GET) falls through to an unconditional create attempt for every generic variable

- 👍 Concur with CodeRabbit.ai feedback - However, I think it's **Normal** severity (not **Major**))

#### Problem

- In `ensure_required_langflow_global_variables`, if the initial `GET /api/v1/variables/` fails or throws (line 130-136), `existing_by_name` stays `{}`. The "create missing generic variables" loop (line 190-197) then treats every entry in `LANGFLOW_GENERIC_GLOBAL_VARIABLES` as missing and issues a `POST` for each.

#### Background Information

- CodeRabbit rated this "Major" and warned it "can create duplicate variables." In practice this is lower risk than that framing suggests: `_upsert_langflow_global_variable` → `_create_langflow_global_variable` (`src/config/settings.py:~1506-1539`) already handles the "already exists" case by falling back to `_update_langflow_global_variable` on a `400`, so a transient GET failure degrades to a wasteful create→400→update round trip rather than an actual duplicate, assuming Langflow enforces per-user name uniqueness (it does — `create_variable` in the vendored Langflow API explicitly checks `variable.name in await variable_service.list_variables(...)` and raises `400` on collision).
- It's also not a regression introduced by this PR: the pre-PR code unconditionally called `_upsert_langflow_global_variable` for every generic variable on every startup with no preceding GET at all, so this fallback path is a strict improvement over prior behavior, just an incomplete one.

#### Code References

- `src/api/settings/langflow_sync.py:129-136, 190-197`

#### Potential Solution

- Return early (after logging) when the listing GET fails, rather than falling into the create loop, and add a test asserting a GET `500` results in zero POSTs — as CodeRabbit suggested.

File: src/api/settings/langflow_sync.py [hunk 4]
```diff
@@ -122,91 +122,142 @@ def _required_generic_global_values(config) -> dict[str, str]:
...
     for name in sorted(LANGFLOW_GENERIC_GLOBAL_VARIABLES):
-        await _upsert_langflow_global_variable(name, _string_value(required_values.get(name, "")))
+        if name not in existing_by_name:
+            try:
+                target_val = _string_value(required_values.get(name, ""))
+                await _upsert_langflow_global_variable(name, target_val)
+            except Exception as e:
+                logger.warning(f"Failed to create Langflow global variable '{name}': {e}")
```
Comment (mpawlow) [inline src/api/settings/langflow_sync.py]: ### (1e) [Minor] New log calls mix raw f-strings with the file's established structured-kwarg logging style

#### Problem

- Several new log lines pass a pre-formatted f-string as the sole message argument (`langflow_sync.py:188, 197, 206, 208`: e.g. `logger.warning(f"Failed to update Langflow global variable '{name}': {e}")`), while adjacent, similarly-new code in the same function uses `structlog`-style structured keyword arguments (lines 155-160, 173-175: `logger.info("Migrating Langflow global variable type", variable_name=name, old_type=curr_type, new_type=target_type)`).

#### Code References

- `src/api/settings/langflow_sync.py:188, 197, 206, 208`

#### Potential Solution

- Convert to structured kwargs, e.g. `logger.warning("Failed to update Langflow global variable", variable_name=name, error=str(e))`, matching the style used elsewhere in the same file/function.

**Review by mpawlow — CHANGES_REQUESTED** [review]
## Functional Review 2

- ✅  Successfully resolves the underlying problem in my workspace 🙏 

## Code Review 1

- See Major PR comments for evaluation: (1a), (1b)
- See Normal PR comment for consideration: (1c)
- See Minor PR comments for optional consideration: (1e)

**Review by mpawlow — APPROVED** [review]
## Code Review 2

- ✅  Approved / LGTM  🚀
